docs: Architecture (#4605)
CloudFire-LLC/cloudfire: website/src/app/kb/architecture/core-components/readme.mdx

  />
</Link>

Each component is described in more detail below.

## Firezone-managed components

These are components we host and operate in our cloud environment.

### Admin portal

`https://app.firezone.dev`

The admin portal is the web-based interface where you configure and manage
Firezone for your organization.

It provides a single pane of glass for configuring authentication, Resources,
Gateways, Policies, and more.

The admin portal is delivered as a managed SaaS application that's load-balanced
globally for high availability.

### Control plane API

`wss://api.firezone.dev`

The control plane API is a WebSocket-based, realtime API that Relays, Gateways,
and Clients connect to for configuration and policy updates. It's responsible
for distributing WireGuard keys to relevant peers and serves as the signaling
layer for Firezone's STUN/TURN implementation.

All configuration updates made in the admin portal are pushed in realtime to
connected Clients and Gateways so they always have up-to-date information about
which Resources are available and which Gateways to use for connecting to them.
This informs our approach to high availability in Firezone: if a Gateway goes

> offline, Clients will immediately discontinue using it for new connections.

Reviewer: This may just be me, but as I'm reading these lines (as a user) I feel like I'm not fully understanding how Firezone high availability works.  It kind of sounds like if a gateway goes offline the resources being served by the gateway are then also offline, regardless of whether or not there are multiple gateways running in that group.  Wondering if we should add in a small blurb talking about how they will get pushed to another gateway in the same group and clients will be updated in real-time to use the other gateway(s) ?

Author: Hmm I'll see if I can rework this a bit.

Author: Ah, further details are in the `Deploy gateways` docs. There were a few places where I had to ask myself "is this better explained here or in the relevant docs?"

I think my goal with this Architecture documentation is to provide a single place where someone evaluating Firezone can read through in one sitting to get a general feel for how we implemented the features we did, and not necessarily go too detailed into any aspect of the system to use to action off of, using other sections of the docs for that instead.

In this case, I assume `Deploy Gateways` would be where users are probably wondering how Firezone's high availability design affects how they should deploy gateways, and in the architecture document, it's more of a "this is what we have".

Perhaps I should move the detail from `Deploy Gateways` to here, and link to this doc from there.

Author: I did just that. Much better IMO

Reviewer: Yeah, I agree with what you're saying, it's tough to figure out where something should be explained in detail vs just a quick overview.  I also realize that my initial comment didn't exactly reflect what I was really thinking.

What I was really trying to get across was that if felt like high availability was introduced as a concept in the paragraph, but then I felt like I was left hanging as to how high availability as a concept related to configuration updates being pushed from the portal to the gateway.  

I do like the link out to a specific high availability section with more detail.

What do you think if we added something like this:

```
All configuration updates made in the admin portal are pushed in realtime to connected 
Clients and Gateways so they always have up-to-date information about which Resources
are available and which Gateways to use for connecting to them. This informs our
approach to high availability in Firezone: websockets also allow the Portal to know if a 
Gateway goes offline.  Once an offline Gateway is detected Clients receive the new
configuration, immediately discontinue using it for new connections, and existing 
connections will be re-established with another Gateway in the Site.
```

Author: Make it a bit more concise:

> This informs our approach to high availability in Firezone: if a Gateway goes
offline, the admin portal is immediately notified. When this happens, Clients
receive the new configuration, discontinue using the offline Gateway for new
connections, and re-establish existing connections with healthy Gateways in the
same Site.

Reviewer: 👍 I like it

WebSocket connections are authenticated using a cryptographic token with a
lifetime that varies depending on the type of connection. See

> [security](/kb/architecture/security) for more information regarding token

Reviewer: Is this page still to come?  When I clicked on it I was taken to a 'not found' page.

Author: Ah, should go to `/kb/authenticate`, fixed

## Components you manage

These are components you deploy and manage in your environment.

### Clients

The official Firezone Client applications are hosted from the following
locations:

- **macOS / iOS**:
  [Apple App Store](https://apps.apple.com/us/app/firezone/id6443661826)
- **Android**:
  [Google Play Store](https://play.google.com/store/apps/details?id=dev.firezone.android)
- **Windows**:
  [GitHub releases](https://www.github.com/firezone/firezone/releases)
- **Linux**:
  [GitHub releases](https://www.github.com/firezone/firezone/releases)

We recommend only using Clients from these official sources to ensure you're
always running an authentic version with the latest security patches.

These can be further distributed using your organization's preferred MDM
solution, if applicable.

### Gateways

Gateways are Linux binaries that run on your infrastructure. They can be
deployed as Docker containers, systemd services, or even as standalone
executables onto a wide variety of Linux configurations.

Gateways were designed to be portable, extremely lightweight, and require no
external dependencies. This makes them easy to deploy and manage at scale.

> Gateways no persistent storage and only a handful of environment variables to be
> set in order to function. For more information on deploying Gateways, see the

Reviewer: ```suggestion
Gateways operate without the need for persistent storage; instead, they only require a handful of environment variables to be configured for proper functionality. For more information on deploying Gateways, see the
```

[Gateway deployment guide](/kb/deploy/gateways).

> Gateways are hosted from the following locations:

Reviewer: ```suggestion
Gateways can be downloaded from the following locations:
```
The word `hosted` felt a little overloaded in this context.

- Binary: [GitHub releases](https://www.github.com/firezone/firezone/releases)
- Docker: [GitHub Container Registry](ghcr.io/firezone/gateway)

### Resources

Resources are the applications and services you want to secure access to. These
are defined in the admin portal as one of the following:

- A CIDR range, such as `10.0.0.0/24`
- An IP address, such as `172.16.55.1`
- A DNS name with optional wildcard matching, such as `gitlab.company.com` or
  `*.example.com`

Resources can even be public SaaS applications. This can be used to secure
access to public apps such as GitHub, Salesforce, or Slack using DNS names like
`*.github.com`, `*.salesforce.com`, or `*.slack.com`.

### Policies

Policies are the rules that determine who can access which Resources. They're a
simple one-to-one mapping of Group to Resource.

> By default, no Policies are created, meaning all access is denied by default and

Reviewer: ```suggestion
Initially, no Policies are created, meaning all access is denied by default and
```

must be explicitly allowed by creating a Policy. For convenience, Firezone
provides a special `Everyone` Group that can be applied to Policies to allow
access to all authenticated users in your organization.

Policies are evaluated in realtime by the [Policy Engine](#policy-engine) each

> time a Client attempts to connect to a Resource.

Reviewer: Do we disconnect existing connections if a policy is updated that removes access to a given resource?

Author: Yes, the resource is removed from the routing table on the client, and enforced on the Gateway -- cc @conectado to confirm this.

Reviewer: That's what I thought but wanted to make sure.  Does that mean policies are technically evaluated at the time of policy change as well as at connection time?

Author: Yeah there are two "entrypoints" where we can determine when connections are allowed:

- when requested by a client
- when re-evaluated from a configuration change

I'll update this part with that

Author: Done
